controlers/usersCtrl.js

```diff
@@ -0,0 +1,97 @@
+const User = require('../models/users.js');
+const Joi = require('joi');
+const jwt = require('jsonwebtoken');
+
+const userJoiSchema = Joi.object({
+    password: Joi.string().min(8).max(20).required(),
+    email: Joi.string().email().required(),
+    subscription: Joi.string().valid('starter', 'pro', 'business').default('starter'),
+    token: Joi.string().allow(null)
+})
+
+const createUser = async (req, res, next) => {
+    const { email, password } = req.body;
+    const user = await User.findOne({email}).lean();
+    const validateResult = userJoiSchema.validate(req.body);
+    if (validateResult.error) {
+        return res.status(400).json({message: 'Błąd z Joi lub innej biblioteki walidacji'})
+    }
+    if (user) {
+        return res.status(409).json({message: 'Email in use'});
+    }
+    try {
+        const newUser = new User({email, password});
+        await newUser.setPassword(password);
+        await newUser.save();
+        return res.status(201).json({message: `User ${req.body.email} created. Subscription: starter`});
+    } catch (err) {
+        next(err)
+    }
+};
+
+const loginUser = async (req, res) => {
+    const { email, password } = req.body;
+    const user = await User.findOne({email});
+    const validateResult = userJoiSchema.validate(req.body);
+    if (validateResult.error) {
+        return res.status(400).json({message: 'Błąd z Joi lub innej biblioteki walidacji'});
+    }
+    if (!user) {
+        return res.status(401).json({message: 'Wrong email'});
+    }
+
+    const isPasswordCorrect = await user.validatePassword(password);
+    if (isPasswordCorrect) {
+        const payload = {
+            id: user._id,
+            email: user.email
+        };
+        const token = jwt.sign(
+            payload,
+            process.env.SECRET,
+            {expiresIn: '60s'}
+        )
+        await user.save();
+        return res.json({token});
+    } else {
+        return res.status(401).json({message: 'Wrong password'});
+    }
+};
+
+const logoutUser = async (req, res, next) => {
+    const userId = req.user._id;
+    try {
+        const user = await User.findById(userId);
+        if (!user) {
+            return res.status(401).json({ message: 'Not authorized' });
+        }
```

Comment:
Jeśli używasz middleware do uwierzytelnienia to jeśli kod dotrze tutaj, to znaczy że już masz usera uwierzytelnionego.

Reply:
Zrobione